imp_developer.py

```diff
@@ -272,56 +326,414 @@ def __base64_encode(str):
                                      data=data,
                                      method=method)
 
         result, code = None, None
         try:
-            res = urllib.request.urlopen(req, timeout=timeout)
+            res = urllib.request.urlopen(req, timeout=None)
             code = res.getcode()
-            result = json.loads(res.read().decode('utf-8'))
+            pl = res.read().decode('utf-8')
+            if pl:
+                result = json.loads(pl)
         except socket.timeout:
             log_debug("Timeout error occurred for URL: " + url)
             result = code
         except urllib.error.HTTPError as err:
             code = err.code
             result = json.loads(err.read().decode('utf-8'))
+        except urllib.error.URLError as err:
+            code = 404
+            result = {"error" : err.reason}
+        except urllib.error.ContentTooShortError as err:
+            code = 404
+            result = {"error" : "Too short conent exception"}
+
         return result, code
 
     @staticmethod
-    def get(key, url, timeout=None):
-        return HTTP.do_request(key, url, "GET", timeout=timeout)
+    def get(key, url, timeout=None, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "GET", timeout=timeout, data=data, headers=headers)
 
     @staticmethod
-    def post(key, url, data=None):
-        return HTTP.do_request(key, url, "POST", data)
+    def post(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "POST", data, headers=headers)
 
     @staticmethod
-    def put(key, url, data=None):
-        return HTTP.do_request(key, url, "PUT", data)
+    def put(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "PUT", data, headers=headers)
+
+    @staticmethod
+    def delete(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "DELETE", data, headers=headers)
 
     @staticmethod
     def is_response_code_valid(code):
-        return code in [200, 201, 202]
+        return code in [200, 201, 202, 204]
+
+    @staticmethod
+    def is_invalid_credentials(code, response):
+        if code in [401]:
+            errors = response.get("errors")
+            if errors and len(errors) > 0:
+                # it is possible to get more than one error
+                # but usually it is only one error
+                for error in errors:
+                    if error.get("title") in ["'Invalid Credentials'"] or error.get("code") == "PX100":
+                        return True
+        return False
+
+    @staticmethod
+    def is_wrong_input(code, response):
+        if code in [400, 409]:
+            errors = response.get("errors")
+            if errors and len(errors) > 0:
+                # it is possible to get more than one error
+                # but usually it is only one error
+                for error in errors:
```

Comment:
why are we dropping the rest of the errors on the floor?

Reply:
Because we need to be sure that we have correct credentials first in this method